src/generateDocumentation.js

```diff
@@ -0,0 +1,311 @@
+'use strict';
+
+const YAML = require('yaml');
+const fileUtils = require('./fileUtils.js');
+
+module.exports = {
+  generateDocumentation: function () {
+    const template = (this.customVars && this.customVars.documentation) ? this.customVars.documentation : {};
+    const version = getOpenApiVersion(this.options.exportType);
+
+    const documentation = {};
+    if (version === "2.0") {
+      documentation.swagger = version;
+    } else {
+      documentation.openapi = version;
+    }
+    documentation.info = generateInfo(template.api);
+    documentation.paths = generatePaths(version, this.serverless.service);
+
+    const models = generateModels(template.models);
+    if (version === "2.0") {
+      if (models && Object.keys(models).length > 0) {
+        documentation.definitions = models;
+      }
+      if (template.securitySchemes) {
+        documentation.securityDefinitions = template.securitySchemes;
+      }
+    } else {
+      if ((models && Object.keys(models).length > 0) || template.securitySchemes) {
+        documentation.components = {
+          schemas: models,
+          securitySchemes: template.securitySchemes || {},
+        };
+      }
+    }
+
+    if (template.api && template.api.tags) {
+      documentation.tags = template.api.tags;
+    }
+
+    const extension = fileUtils.getFileExtension(this.options.outputFileName);
+    let output;
+    if (fileUtils.isYamlFileExtension(extension)) {
+      output = YAML.stringify(documentation, { indent: 2 });
+    } else {
+      output = JSON.stringify(documentation, null, 2);
+    }
+    this.fs.writeFile(this.options.outputFileName, output);
```

Comment:
@rovellipaolo - Now I can see the change. Perhaps it was just a disturbance in the force.

Although.. I have seconds thoughts about this... If I understand the code well this function will actually actually generate and download the documentation. If we keep this async... that means that the command exists before the document actually gets downloaded?

Reply:
> If I understand the code well this function will actually actually generate and download the documentation.

This function generates the documentation locally and writes/stores it into a file.

> If we keep this async... that means that the command exists before the document actually gets downloaded?

No. I don't really know the internals of the Serverless framework and plugins system (this is the very first Serverless plugin I touch), but I tried it and apparently the command waits until the file is written.

You can see this, for example, by applying the following diff:
```
diff --git a/src/generateDocumentation.js b/src/generateDocumentation.js
index 0be8d04..13904a1 100644
--- a/src/generateDocumentation.js
+++ b/src/generateDocumentation.js
@@ -45,6 +45,13 @@ module.exports = {
     } else {
       output = JSON.stringify(documentation, null, 2);
     }
+    new Promise((resolve) => {
+        process.stdout.write("Start");
+        setTimeout(() => {
+            resolve();
+            process.stdout.write("Stop");
+        }, 10000)
+    });
     this.fs.writeFile(this.options.outputFileName, output);
   },
 };
 ```
 And then running: `serverless generateDocumentation --outputFileName openapi.json --exportType oas30`

Besides, as far as I can see, also the `downloadDocumentation` command is writing a file inside a Promise: https://github.com/failsafe-engineering/serverless-aws-apigateway-documentation/blob/8be76400ac7a926e2583cd7935a4f7e653712c4d/src/downloadDocumentation.js#L4-L20

Reply:
This issues is quite old but, according to what written here, if the hook function returns a Promise then handler will be awaited: https://github.com/serverless/serverless/issues/4557

I've added an explicit return for clearness: https://github.com/failsafe-engineering/serverless-aws-apigateway-documentation/pull/43/commits/74e8c282f8ae26d8f7ff7ad2bf85bf8cc2f1ba9c
That said, if you prefer me to revert it to write the output file synchronously, no problem.

Reply:
@rovellipaolo - Thanks for clarifying it. Approved. I'll create a follow up PR and release the change as a minor.